### PR TITLE
Implement core abstract float support

### LINKED ISSUE #2416: Add AbstractFloat support to floating point testing framework
This covers all of the refactoring of the existing code to support parameterization on floating point type, and related work to get that working. This does not cover actually enabling specific AbstractFloat tests, those are covered by other issues.

### PR DESCRIPTION
**Pull request by zoddicus**
This PR adds in the various utilities needed for abstract float (f64), i.e. nextAfter, ULP, serialize/deserialize interval.

An implementation FPTraits, FPAbstractTraits is implemented to use these utiliites.

Test coverage is added for all of this new code.

The interval generators are roughed in enough to instantiate the FPTraits object, but not tested or used. Further PRs will implement testing for this code and add it being used.

Fixes #2416

<hr>

**Requirements for PR author:**

- [x] All missing test coverage is tracked with "TODO" or `.unimplemented()`.
- [x] New helpers are `/** documented */` and new helper files are found in `helper_index.txt`.
- [x] Test behaves as expected in a WebGPU implementation. (If not passing, explain above.)

**Requirements for [reviewer sign-off](https://github.com/gpuweb/cts/blob/main/docs/reviews.md):**

- [ ] Tests are properly located in the test tree.
- [ ] [Test descriptions](https://github.com/gpuweb/cts/blob/main/docs/intro/plans.md) allow a reader to "read only the test plans and evaluate coverage completeness", and accurately reflect the test code.
- [ ] Tests provide complete coverage (including validation control cases). **Missing coverage MUST be covered by TODOs.**
- [ ] Helpers and types promote readability and maintainability.

When landing this PR, be sure to make any necessary issue status updates.

### CODE REVIEW

**Review by ben-clayton — APPROVED** [review]
Looks great - well done!

It's an enormous amount of code, and I've tried by best to check what I can, but would be sensible to get a second approval before landing.

File: src/webgpu/util/constants.ts
Comment (zoddicus) [reply src/webgpu/util/constants.ts:439]: Created a tracking issue, https://github.com/gpuweb/cts/issues/2500.

I agree with changing the names, they originated as helpers to convert a integer value written out in hex into a floating point value, so I could specify the bits, but have definitely outgrown this.

File: src/unittests/maths.spec.ts
Comment (dneto0) [inline src/unittests/maths.spec.ts:577]: Manually verified this is in the right ballpark.

File: src/webgpu/util/constants.ts
Comment (dneto0) [inline src/webgpu/util/constants.ts:438]: For a followup CL.  It's not entirely clear what this constant is trying to capture.

Is it the number that is just outside the finite range of f32 such that it's the smallest f64 that would convert to an f32 infinity?   I guess I'm not clear on the derivation.

File: src/webgpu/util/floating_point.ts
Comment (dneto0) [inline src/webgpu/util/floating_point.ts:4464]: Nice reuse!

File: src/webgpu/util/math.ts
Comment (dneto0) [inline src/webgpu/util/math.ts:351]: Ack.

**Review by dneto0 — APPROVED** [review]
Thanks!
Really nice detailed work.
Great reuse.

**Review by dneto0 — APPROVED** [review]

File: src/webgpu/util/constants.ts
Comment (ben-clayton) [reply src/webgpu/util/constants.ts]: > Quiet and signalling nan differ in the top bit of the significand.

That's an implementation detail. This is not specified by any specification that I'm aware of. We shouldn't be relying on it.

Comment (dneto0) [reply src/webgpu/util/constants.ts]: It's something I remember "forever", so I checked IEEE 754 2008.  Definitely something I didn't read from that spec before, but it says:

> 6.2.1 NaN encodings in binary formats

> ....  A quiet NaN bit
string should be encoded with the first bit (d1) of the trailing significand field T being 1. A signaling NaN
bit string should be encoded with the first bit of the trailing significand field being 0. If the first bit of the
trailing significand field is 0, some other bit of the trailing significand field must be non-zero to distinguish
the NaN from infinity. 

In an earlier diagram d1 is shown as the MSB of the significand.

It also generally says that the remaining bits of the significand should encode diagnostic information, and that NaNs should be passed on through operations, to propagate that diagnostic info.


https://en.wikipedia.org/wiki/NaN#Encoding says that many processors follow this convention.
Amusingly, PA-RISC and old MIPS CPUs flip the sense of that top bit.  It also gives a rationale for following IEEE (you can convert signalling to quiet by blindly setting the top bit to 1.)

All that said, WGSL does not specify this detail.  So we do have to be careful about overconstraining the tests.

What this *is* good for is to introduce diversity in the test inputs, because it's a likely thing to trigger differing behaviour, in the CPU side, and in GPUs that try harder to do things the IEEE way.

Comment (ben-clayton) [reply src/webgpu/util/constants.ts]: Humm, I take it back. I had read previously that this signalling bit was undefined. Clearly that was wrong. My apologies.

**Pull request merged**
